### PR TITLE
Add security considerations for targets connecting to listener

### PR DESCRIPTION
**Pull request by asingh-g**
~Leave it up to the implementation to determine if they should restrict connecting IPs or allow the proxy to act as a server~

EDIT: As discussed below, allow all targets to be forwarded, add security consideration about this

### CODE REVIEW

File: draft-ietf-masque-connect-udp-listen.md [hunk 1]
```diff
@@ -188,6 +188,10 @@ validate the target on every datagram and MUST NOT forward individual datagrams
 with unauthorized targets. Proxies can either silently discard such datagrams or
 abort the corresponding request stream.
 
+In addition, Proxies SHOULD consider restricting the IPs allowed to connect
+to the client via the proxy and reject any unauthorized connections on the
+target side, in order to prevent the misuse of the proxy.
```
Comment (marten-seemann) [inline draft-ietf-masque-connect-udp-listen.md:193]: It is not clear to me how the proxy would know which IPs are safe and which ones aren't.

**Review by marten-seemann — CHANGES_REQUESTED** [review]
I don't believe this resolves #13. That issue is asking for introducing a way for clients to request proxies to only allow incoming packets from certain targets.

File: draft-ietf-masque-connect-udp-listen.md (hunk 1 above)
Comment (asingh-g) [reply draft-ietf-masque-connect-udp-listen.md:193]: Determining what qualifies as a safe or accepted target is an exercise that should be left to the proxy implementation in my opinion. This applies both ways, the proxy may only want to provide its proxying capability to certain targets, and likewise only let certain targets talk to the client, or they allow all traffic either way, knowing the risks involved, up to them to monitor or limit the traffic.

**Comment by asingh-g** [comment]
> I don't believe this resolves #13. That issue is asking for introducing a way for clients to request proxies to only allow incoming packets from certain targets.


The client could simply try to reach those targets via the proxy, effectively creating the permission for it. I believe that this is similar to how networks firewalls function and how WebRTC gets around these issues.

**Comment by asingh-g** [comment]
And to add to my previous comment, the cost here is one UDP packet being sent out per target it wants to receive from. Which doesn't seem significant, unless we want the client to be able to open permissions for entire subnets. At that point, it acts as a server.
Thoughts @DavidSchinazi ?

File: draft-ietf-masque-connect-udp-listen.md (hunk 1 above)
Comment (DavidSchinazi) [reply draft-ietf-masque-connect-udp-listen.md:193]: Unfortunately, leaving this as exercise to the proxy leads to interoperability issues. In this case, one proxy might decide to restrict to addresses that the client has sent to, whereas another might restrict to address and ports. Now clients don't know which is which and have to guess. It's almost always better to be explicit when it comes to load-bearing behavior like this.

I would instead mention in security considerations that, unlike in regular CONNECT-UDP, the clients can now receive unwanted traffic from other hosts, and should be ready to handle it.

**Comment by marten-seemann** [comment]
> > I don't believe this resolves #13. That issue is asking for introducing a way for clients to request proxies to only allow incoming packets from certain targets.
> 
> The client could simply try to reach those targets via the proxy, effectively creating the permission for them. I believe that this is similar to how network firewalls function and how WebRTC gets around these issues.

This would be very unfortunate, and jeopardize entire use cases for this draft. At the very least there should be explicit config option that needs to be set by the client. Having to enable each incoming address separately would defeat the purpose of creating a publicly reachable listener, and make it impossible to use it for the initial proxied QUIC connections in the [QUIC NAT Traversal draft](https://datatracker.ietf.org/doc/draft-seemann-quic-nat-traversal/).

**Comment by asingh-g** [comment]
Taking @DavidSchinazi 's suggestion, I've changed the security consideration. We leave it default open to forwarding all traffic from any targets to the client. Let me know what you think of this proposal @marten-seemann.

File: draft-ietf-masque-connect-udp-listen.md [hunk 2]
```diff
@@ -188,6 +188,10 @@ validate the target on every datagram and MUST NOT forward individual datagrams
 with unauthorized targets. Proxies can either silently discard such datagrams or
 abort the corresponding request stream.
 
+Unlike non binding CONNECT-UDP, since the proxies tunnel datagrams from any
```
Comment (marten-seemann) [inline draft-ietf-masque-connect-udp-listen.md:191]:
```suggestion
Unlike non-binding CONNECT-UDP, since the proxy tunnels datagrams from any
```

**Review by marten-seemann — APPROVED** [review]

**Comment by marten-seemann** [comment]
I still don't think that this PR addresses #13 though. This issue should not be closed when this PR is merged.

**Comment by marten-seemann** [comment]
I'm not disagreeing here. All I'm saying is that this PR does *not* resolve #13 (as it currently claims it does). It might or might not be fine to not resolve #13, but why not discuss that on the issue?

**Comment by asingh-g** [comment]
Okay that's valid, I'll add the some comments there

File: draft-ietf-masque-connect-udp-listen.md [hunk 3]
```diff
@@ -188,6 +188,10 @@ validate the target on every datagram and MUST NOT forward individual datagrams
...
+target to clients bound to their respective public IP and ports, the clients
+SHOULD be ready to handle potential unwanted traffic from unknown destinations.
```
Comment (DavidSchinazi) [inline draft-ietf-masque-connect-udp-listen.md:193]: Given the security implications, this needs to be mandatory
```suggestion
MUST be ready to handle potential unwanted traffic from unknown destinations.
```

**Review by DavidSchinazi — APPROVED** [review]

**Pull request merged**